packages/@lwc/template-compiler/src/parser/expression-complex/html.ts

```diff
@@ -146,78 +162,97 @@ class TemplateHtmlTokenizer extends Tokenizer {
-            // Moving the cursor back by one allows the state machine to correctly detect
-            // the state into which it should next transition.
-            this._unconsume();
+
+            this._advanceBy(this.currentAttr!.value.length - 1);
+            this.consumedAfterSnapshot = this.currentAttr!.value.length;
         } else {
             // If the first character in an unquoted-attr-value is not an opening
             // curly brace, it isn't a template expression. Opening curly braces
             // coming later in an unquoted attr value should not be considered
             // the beginning of a template expression.
             this.checkedAttrs.add(this.currentAttr);
-            super.ATTRIBUTE_VALUE_UNQUOTED_STATE(codePoint);
+            // @ts-ignore
+            super._stateAttributeValueUnquoted(codePoint);
         }
     }
 
     // DATA_STATE is the initial & default state of the lexer. It can be thought of as the
     // state when the cursor is outside of an (opening or closing) tag, and outside of
     // special parts of an HTML document like the contents of a <style> or <script> tag.
     // In other words, we're parsing a text node when in DATA_STATE.
-    DATA_STATE(codePoint: number) {
+    _stateData(codePoint: number) {
         if (codePoint === OPENING_CURLY_BRACKET) {
             // An opening curly brace may be the first character in a text node.
             // If that is not the case, we need to emit the text node characters
             // that come before the curly brace.
             if (this.currentCharacterToken) {
+                this.currentLocation = this.getCurrentLocation(0);
+                this.consumedAfterSnapshot = 0;
                 // Emit the text segment preceding the curly brace.
-                this._emitCurrentCharacterToken();
+                this._emitCurrentCharacterToken(this.currentLocation);
             }
 
             const expressionTextNodeValue = this.parseTemplateExpression();
 
             // Create a new text-node token to contain our `{expression}`
-            this._createCharacterToken(Tokenizer.CHARACTER_TOKEN, expressionTextNodeValue);
+            this._createCharacterToken(Token.TokenType.CHARACTER, expressionTextNodeValue);
+
+            this._advanceBy(expressionTextNodeValue.length);
+            this.currentLocation = this.getCurrentLocation(0);
 
             // Emit the text node token containing the `{expression}`
-            this._emitCurrentCharacterToken();
+            this._emitCurrentCharacterToken(this.currentLocation);
 
             // Moving the cursor back by one allows the state machine to correctly detect
             // the state into which it should next transition.
-            this._unconsume();
+            this.preprocessor.retreat(1);
+
+            this.currentToken = null;
+            this.currentCharacterToken = null;
         } else {
-            super.DATA_STATE(codePoint);
+            // @ts-ignore
```

Comment:
Add a tracking issue for upstream `private` to `protected`.

Reply:
This issue was fixed in https://github.com/inikulin/parse5/pull/996. I can come back in a second PR and remove the type workarounds that are now redundant.